SonarScanner.Shim/PropertiesFileGenerator.cs

```diff
@@ -102,48 +102,125 @@ public bool TryWriteProperties(PropertiesWriter writer, out IEnumerable<ProjectD
 
             FixSarifAndEncoding(projects, analysisProperties);
 
-            var rootProjectBaseDir = ComputeRootProjectBaseDir(projectPaths);
-
             allProjects = projects
                 .GroupBy(p => p.ProjectGuid)
-                .Select(g => ToProjectData(g, rootProjectBaseDir))
+                .Select(ToProjectData)
                 .ToList();
 
             var validProjects = allProjects
                 .Where(d => d.Status == ProjectInfoValidity.Valid)
                 .ToList();
 
-            if (validProjects.Any())
+            if (validProjects.Count == 0)
+            {
+                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                return false;
+            }
+
+            var rootProjectBaseDir = ComputeRootProjectBaseDir(projectPaths);
+            var rootModuleFiles = PutFilesToRightModuleOrRoot(validProjects, rootProjectBaseDir).ToList();
+            PostProcessProjectStatus(validProjects);
+
+            if (rootModuleFiles.Count == 0 && validProjects.All(p => p.Status == ProjectInfoValidity.NoFilesToAnalyze))
             {
-                var sharedFilePaths = validProjects
-                    .SelectMany(p => p.ExternalFiles)
-                    .Where(f => !PathHelper.IsPartOfAProject(f, projectPaths))
-                    .Distinct(StringComparer.InvariantCultureIgnoreCase)
-                    .ToList();
+                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                return false;
+            }
 
-                writer.WriteSonarProjectInfo(rootProjectBaseDir);
-                writer.WriteSharedFiles(sharedFilePaths);
+            writer.WriteSonarProjectInfo(rootProjectBaseDir);
+            writer.WriteSharedFiles(rootModuleFiles);
+            validProjects.ForEach(p => writer.WriteSettingsForProject(p, p.CoverageAnalysisExists(logger)
+                ? p.VisualStudioCoverageLocation : null));
 
-                foreach (var projectData in validProjects)
+            // Handle global settings
+            writer.WriteGlobalSettings(analysisProperties);
+
+            return true;
+        }
+
+        private IEnumerable<string> PutFilesToRightModuleOrRoot(IEnumerable<ProjectData> projects, string rootProjectBaseDir)
```

Comment:
A brief comment would help with maintenance in the future; the problem being solved, and what the return value is (I guess it's the list of files allocated to the root?)